### ISSUE TITLE
Container does not support changing the server.baseUrl configuration

### ISSUE THREAD

**Issue by jerryharrow** [bitnami/bitnami-docker-kibana#48]
Attempting to change the server.basePathl (in my case using the bitnami helm chart configuration.server.basePath) and the container does not start.

**Steps to reproduce the issue:**

1.  Deploy kibana using the helm chart with  configuration.server.basePath specified, or by specifying the server.basePath in the kibana.yml

2. Check container log and you will see the container failing to start with no error message.

$ kubectl logs kibana:
```
kibana 18:33:34.77
kibana 18:33:34.77 Welcome to the Bitnami kibana container
kibana 18:33:34.78 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-kibana
kibana 18:33:34.78 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-kibana/issues
kibana 18:33:34.78
kibana 18:33:34.78 INFO  ==> ** Starting Kibana setup **
kibana 18:33:34.81 INFO  ==> Configuring/Initializing Kibana...
kibana 18:33:34.86 INFO  ==> Found mounted configuration directory
kibana 18:33:34.87 INFO  ==> Starting Kibana in background
kibana 18:33:34.88 INFO  ==> Waiting for Kibana to be started and ready
kibana 18:34:37.11 ERROR ==> Kibana is not available after 30 retries
```

Should have started at the specified configuration.server.basePath

**Version**

bitnami/kibana:7.12.1-debian-10-r0

It appears that the problem is that the is_kibana_ready() method in libkibana.sh
https://github.com/bitnami/bitnami-docker-kibana/blob/05d237435b036a577cec50fbce990c3520d61982/7.10.2/debian-10/rootfs/opt/bitnami/scripts/libkibana.sh#L185

Is unconditionally polling (curl -s "127.0.0.1:${KIBANA_PORT_NUMBER}/api/status")")" -- but needs to include server.basePath value.

**Comment by migruiz4** [comment]
Hi @jerryharrow,

I haven't been able to reproduce your issue and Kibana starts properly using: 

```
configuration:
  server:
    basePath: "/kibanatest"
    rewriteBasePath: false
```

Could you please try adding the following extra envVars to your deployment so we get some additional information from Kibana?
```
extraEnvVars:
  - name: BITNAMI_DEBUG
    value: "true"
```

**Comment by jerryharrow** [comment]
Hi @migruiz4, 

Let me add some details to clarify.   I'm trying to enable kibana for developer use on my existing ngnix ingress shared with the rest of my application.  If I just naively do so it listens on / and hides all my other REST apis -- so I needed to change that ingress path such that the external path is distinct (like /kibanatest as you have above).   Initially, I didn't specify rewriteBasePath and it defaulted to false (despite the documentation -- "Specifies whether Kibana should rewrite requests that are prefixed with server.basePath or require that they are rewritten by your reverse proxy. This setting was effectively always false before Kibana 6.3 and will default to true starting in Kibana 7.0." ).

Anyway it generated:

ovad-kibana-6ff95f9bf4-dg9zr:/opt/bitnami/kibana/config$ more kibana.yml
pid.file: /opt/bitnami/kibana/tmp/kibana.pid
server.host: "::"
server.port: 5601
elasticsearch.hosts: [http://ovad-coordinating-only:9200]
server.basePath: "/kibana"
server.rewriteBasePath: false
kibana.defaultAppId: dashboard/1a16fba0-b978-11eb-8948-a52eb9aa5677

But it still  didn't start successfully -- as the "127.0.0.1:${KIBANA_PORT_NUMBER}/api/status" is actually at "127.0.0.1:${KIBANA_PORT_NUMBER}/kibana/api/status"), with server.basePath: "/kibana", right?

kibana 15:02:51.35
kibana 15:02:51.36 Welcome to the Bitnami kibana container
kibana 15:02:51.36 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-kibana
kibana 15:02:51.37 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-kibana/issues
kibana 15:02:51.37
kibana 15:02:51.38 INFO  ==> ** Starting Kibana setup **
kibana 15:02:51.39 DEBUG ==> Validating settings in KIBANA_* environment variables...
kibana 15:02:51.41 INFO  ==> Configuring/Initializing Kibana...
kibana 15:02:51.42 DEBUG ==> Ensuring expected directories/files exist...
kibana 15:02:51.47 INFO  ==> Found mounted configuration directory
kibana 15:02:51.49 INFO  ==> Starting Kibana in background
kibana 15:02:51.49 INFO  ==> Waiting for Kibana to be started and ready
kibana 15:02:51.50 DEBUG ==> Waiting for Kibana server: 30 remaining attempts...
kibana 15:02:53.55 DEBUG ==> Waiting for Kibana server: 29 remaining attempts...
kibana 15:02:55.60 DEBUG ==> Waiting for Kibana server: 28 remaining attempts...
kibana 15:02:57.75 DEBUG ==> Waiting for Kibana server: 27 remaining attempts...
kibana 15:02:59.80 DEBUG ==> Waiting for Kibana server: 26 remaining attempts...
kibana 15:03:01.85 DEBUG ==> Waiting for Kibana server: 25 remaining attempts...
kibana 15:03:03.90 DEBUG ==> Waiting for Kibana server: 24 remaining attempts...
kibana 15:03:05.95 DEBUG ==> Waiting for Kibana server: 23 remaining attempts...
kibana 15:03:08.00 DEBUG ==> Waiting for Kibana server: 22 remaining attempts...
kibana 15:03:10.05 DEBUG ==> Waiting for Kibana server: 21 remaining attempts...
kibana 15:03:12.10 DEBUG ==> Waiting for Kibana server: 20 remaining attempts...
kibana 15:03:14.15 DEBUG ==> Waiting for Kibana server: 19 remaining attempts...
kibana 15:03:16.20 DEBUG ==> Waiting for Kibana server: 18 remaining attempts...
kibana 15:03:18.25 DEBUG ==> Waiting for Kibana server: 17 remaining attempts...
kibana 15:03:20.30 DEBUG ==> Waiting for Kibana server: 16 remaining attempts...
kibana 15:03:22.35 DEBUG ==> Waiting for Kibana server: 15 remaining attempts...
kibana 15:03:24.40 DEBUG ==> Waiting for Kibana server: 14 remaining attempts...
kibana 15:03:26.45 DEBUG ==> Waiting for Kibana server: 13 remaining attempts...
kibana 15:03:28.50 DEBUG ==> Waiting for Kibana server: 12 remaining attempts...
kibana 15:03:30.55 DEBUG ==> Waiting for Kibana server: 11 remaining attempts...
kibana 15:03:32.60 DEBUG ==> Waiting for Kibana server: 10 remaining attempts...
kibana 15:03:34.65 DEBUG ==> Waiting for Kibana server: 9 remaining attempts...
kibana 15:03:36.70 DEBUG ==> Waiting for Kibana server: 8 remaining attempts...
kibana 15:03:38.75 DEBUG ==> Waiting for Kibana server: 7 remaining attempts...
kibana 15:03:40.80 DEBUG ==> Waiting for Kibana server: 6 remaining attempts...
kibana 15:03:42.85 DEBUG ==> Waiting for Kibana server: 5 remaining attempts...
kibana 15:03:44.89 DEBUG ==> Waiting for Kibana server: 4 remaining attempts...
kibana 15:03:46.95 DEBUG ==> Waiting for Kibana server: 3 remaining attempts...
kibana 15:03:48.99 DEBUG ==> Waiting for Kibana server: 2 remaining attempts...
kibana 15:03:51.04 DEBUG ==> Waiting for Kibana server: 1 remaining attempts...
kibana 15:03:53.09 ERROR ==> Kibana is not available after 30 retries

During this time, kibana is of course running fine, the is_kibana_ready() is just returning false because the API is not at /api/status anymore.   After applying the change in #49 -- the container would not continually restart because the check is_kibana_ready() succeeded as it should.   Sometimes I could connect to the kibana API during that 30 retries above.

Running in this state the UI still would not function properly because the paths generated by kibana were still relative to / and thus any internal redirections that were not relative to the current URL would fail (there seemed to be  afew of those), so then I added 

But in doing so, the UI still does not work via the ingress because the generated paths in the UI do not add the /kibanatest and thus do not end up getting routed to kibana but are relative to / on the nginx ingress.       So I added an explicit
server.rewriteBasePath: true and in that state, I could then access kibana UI -- either directly through the ingress /kibana -- or using a port forward:

kubectl --kubeconfig kube.conf port-forward --address localhost service/ovad-kibana 8080:5601

**Comment by migruiz4** [comment]
Hi @jerryharrow,

Thank you for the detailed explanation.

If I understood it correctly from Kibana documentation and tests performed locally:

- Using `basePath` and `rewriteBasePath: false`, Kibana REST endpoints continue to reply at `/` as it expects the proxy to rewrite the requests, but the links in the UI will use the custom basePath.
```
> curl http://localhost:8080/api/status        
{"name":"kib-kibana-86c54fff8f-qrxj8","uuid":"384805cd-2fbf-4e3b-8afa-a7c10081d1a5","version":{"n ...
> curl http://localhost:8080/testing/api/status
{"statusCode":404,"error":"Not Found","message":"Not Found"}        
```
- Using `basePath` and `rewriteBasePath:true`, Kibana will move all of its endpoints to the new custom basePath, so the proxy does not need to do a rewrite from `/kibanatest -> /`.
```
> curl http://localhost:8080/testing/api/status
{"name":"kib-kibana-7fb4c86dc9-s4kw6","uuid":"7072376b-3793-47ba-9359-a197f382f643","version":{"n...                                                                                                     
> curl http://localhost:8080/api/status        
{"statusCode":404,"error":"Not Found","message":"Not Found"}       
```

Indeed, there is an issue with this line as you described:
```
 local -r state="$(yq r - "status.overall.state" <<<"$(curl -s "127.0.0.1:${KIBANA_PORT_NUMBER}/api/status")")" 
```
But it is only required when using `server.rewriteBasePath: true` in the Kibana configuration.

**Comment by jerryharrow** [comment]
Hi @migruiz4 

Clearly I'm missing something here.     What does server.basePath do if it doesn't change what URL you pass into the REST interface?      If change the ingress completely via the rewrite rule, then I wouldn't even need server.basePath, right?

I did initially start by attempting to use an ingress rewrite rule, but didn't succeed getting the UI to full work with that.    That would of course be sufficient if I could get it to work fully.    This is what I had previously tried as per https://kubernetes.github.io/ingress-nginx/examples/rewrite.

elasticsearch:
  fullnameOverride: elasticsearch
  global:
    kibanaEnabled: true
    storageClass: "managed-nfs-storage-delete"
  kibana:
    ingress:
      # Need to somehow remove the hostname the chart configures as it always ends up as "kibana.local"
      # but ngnix has no hostname so it does not match.   Hostname "*" is not accepted so no way
      # to get this to match the external controller without a fix to the chart.
      # hostname: "*"
      enabled: true
      path: /kibana(/|$)(.*)
      annotations:
        kubernetes.io/ingress.class: *ingressClass
        nginx.ingress.kubernetes.io/rewrite-target: /$2
    savedObjects:
      configmap: kibana-config
    extraEnvVars:
      - name: BITNAMI_DEBUG
        value: "true"
    extraConfiguration:
      # Dashboard Id from files/kibana-config/export.json
      kibana.defaultAppId: "dashboard/1a16fba0-b978-11eb-8948-a52eb9aa5677"


Any hints as to what I have wrong?

**Comment by migruiz4** [comment]
Hi @jerryharrow,

>Clearly I'm missing something here. What does server.basePath do if it doesn't change what URL you pass into the REST interface? If change the ingress completely via the rewrite rule, then I wouldn't even need server.basePath, right?

When using `basePath` and `rewriteBasePath: false`, Kibana will change all the links in the UI so you don't receive a 404 response every time you click on a link, because the links will point to `/` but your proxy expects them at `/kibanatest/`.

Trying to explain it in a 'diagram' way: 

Browser UI -->/kibanatest --> Nginx --> / --> Kibana

So, by setting `basePath`, Kibana links in the UI will also aim to `/kibanatest`.
To summarize things, you would need both the rewrite rule in Nginx AND the `basePath`+`rewriteBasePath: false`

Obviously, it is much simpler to set `rewriteBasePath: true` and that might be the reason why it is the new default value, but I'd like the container logic to be backward compatible and allow both cases.

**Comment by jerryharrow** [comment]
Fixed by #49
